### PR TITLE
*: support password validation options and variables

### LINKED ISSUE #38928: Support Password Validation Options and Variables
https://dev.mysql.com/doc/mysql-security-excerpt/8.0/en/validate-password-options-variables.html

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #38928 #38924

Problem Summary:

We need to improve the compatibility with MySQL in password validation.

### What is changed and how it works?

Most of the feature is the same as MySQL's, except that: 
1. Since TiDB does not have a mechanism like *components* in MySQL, we add a new global variable `validate_password.enable` to turn on/off the validation.
2. We introduce `validate_password.dictionary` instead of `validate_password.dictionary_file`, since it is hard to upload the dictionary file to the tidb-servers's local file systems.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
1. support password validation when `create user`, `alter user` and `set password`
3. introduce system variables about password validation
4. support system function `validate_password_strength`
```

### CODE REVIEW

File: executor/simple.go [hunk 1]
```diff
@@ -61,6 +54,10 @@ import (
 	"github.com/pingcap/tipb/go-tipb"
 	tikvutil "github.com/tikv/client-go/v2/util"
 	"go.uber.org/zap"
+	"os"
```
Comment (wjhuang2016) [inline executor/simple.go:57]: fmt

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ import (
 	"crypto/sha512"
 	"encoding/binary"
 	"fmt"
+	validatePwd "github.com/pingcap/tidb/util/validate-password"
```
Comment (wjhuang2016) [inline expression/builtin_encryption.go:28]: fmt

File: expression/builtin_encryption.go [hunk 3]
```diff
@@ -1010,5 +1012,64 @@ type validatePasswordStrengthFunctionClass struct {
...
 func (c *validatePasswordStrengthFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "VALIDATE_PASSWORD_STRENGTH")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	charset, collate := ctx.GetSessionVars().GetCharsetInfo()
```
Comment (wjhuang2016) [inline expression/builtin_encryption.go:1022]: The result is int, so you should set the charset and collate.

File: expression/builtin_encryption.go [hunk 4]
```diff
@@ -1010,5 +1012,64 @@ type validatePasswordStrengthFunctionClass struct {
...
+	bf.tp.SetCharset(charset)
+	bf.tp.SetCollate(collate)
+	bf.tp.SetFlen(args[0].GetType().GetFlen())
```
Comment (wjhuang2016) [inline expression/builtin_encryption.go:1025]: Flen should be 21

File: expression/builtin_encryption.go [hunk 5]
```diff
@@ -1010,5 +1012,65 @@ type validatePasswordStrengthFunctionClass struct {
...
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(21)
+	sig := &builtinValidatePasswordStrengthSig{bf}
+	//sig.setPbCode(tipb.ScalarFuncSig_ValidatePasswordStrength)
```
Comment (wjhuang2016) [inline expression/builtin_encryption.go:1024]: ...

File: sessionctx/variable/sysvar.go [hunk 6]
```diff
@@ -486,6 +486,46 @@ var defaultSysVars = []*SysVar{
...
+			} else if length < numberCount+specialCharCount+2*mixedCaseCount {
+				return "", ErrWrongValueForVar.GenWithStackByArgs(ValidatePasswordLength, normalizedValue)
+			}
+			return normalizedValue, nil
+		},
+	},
+	{Scope: ScopeGlobal, Name: ValidatePasswordMixedCaseCount, Value: "1", Type: TypeInt, MinValue: 0, MaxValue: math.MaxInt64},
+	{Scope: ScopeGlobal, Name: ValidatePasswordNumberCount, Value: "1", Type: TypeInt, MinValue: 0, MaxValue: math.MaxInt64},
+	{Scope: ScopeGlobal, Name: ValidatePasswordSpecialCharCount, Value: "1", Type: TypeInt, MinValue: 0, MaxValue: math.MaxInt64},
+	{Scope: ScopeGlobal, Name: ValidatePasswordDictionaryFile, Value: "", Type: TypeStr, SetGlobal: func(_ context.Context, s *SessionVars, val string) error {
```
Comment (wjhuang2016) [inline sessionctx/variable/sysvar.go:521]: How to use it? The path is on the server side. It's hard to pass the file to k8s, DBASS.

File: expression/builtin_encryption.go (hunk 5 above)
Comment (CbcWestwolf) [reply expression/builtin_encryption.go]: Remove :-)

**Review by wjhuang2016 — APPROVED** [review]

File: executor/simple.go [hunk 7]
```diff
@@ -783,6 +784,26 @@ func (e *SimpleExec) executeRollback(s *ast.RollbackStmt) error {
...
+func (e *SimpleExec) authUsingCleartextPwd(authOpt *ast.AuthOption, authPlugin string) bool {
+	if authOpt == nil || !authOpt.ByAuthString {
+		return false
+	}
+	if authPlugin == mysql.AuthNativePassword ||
+		authPlugin == mysql.AuthTiDBSM3Password ||
+		authPlugin == mysql.AuthCachingSha2Password {
+		return true
+	}
+	return false
```
Comment (xiongjiwei) [inline executor/simple.go:796]:
```suggestion
	return authPlugin == mysql.AuthNativePassword ||
		authPlugin == mysql.AuthTiDBSM3Password ||
		authPlugin == mysql.AuthCachingSha2Password
```

File: executor/simple.go [hunk 8]
```diff
@@ -783,6 +784,26 @@ func (e *SimpleExec) executeRollback(s *ast.RollbackStmt) error {
...
+}
+
+func (e *SimpleExec) enableValidatePassword() bool {
```
Comment (xiongjiwei) [inline executor/simple.go:799]:
```suggestion
func (e *SimpleExec) isValidatePasswordenabled() bool {
```

File: util/password-validation/password_validation.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+const minPwdValidationLength int = 4
+
+// ValidateDictionaryPassword checks if the password contains words in the dictionary.
+func ValidateDictionaryPassword(pwd string, globalVars *variable.GlobalVarAccessor) (bool, error) {
+	dictionary, err := (*globalVars).GetGlobalSysVar(variable.ValidatePasswordDictionary)
+	pwdLength := len(pwd)
+	if err != nil {
+		return false, err
+	}
```
Comment (xiongjiwei) [inline util/password-validation/password_validation.go:39]:
```suggestion
	if err != nil {
		return false, err
	}
	pwdLength := len(pwd)
```

File: util/password-validation/password_validation.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	words := strings.Split(dictionary, ";")
+	if len(words) == 0 {
+		return true, nil
+	}
+	cache := make(map[string]interface{}, len(words))
```
Comment (xiongjiwei) [inline util/password-validation/password_validation.go:44]:
```suggestion
	cache := make(map[string]struct{}, len(words))
```

File: util/password-validation/password_validation.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+	for _, word := range words {
+		word = strings.ToLower(word)
+		if len(word) >= minPwdValidationLength && len(word) <= maxPwdValidationLength {
+			cache[word] = nil
```
Comment (xiongjiwei) [inline util/password-validation/password_validation.go:48]:
```suggestion
			cache[word] = struct{}{}
```

File: sessionctx/variable/sysvar.go [hunk 12]
```diff
@@ -486,6 +486,78 @@ var defaultSysVars = []*SysVar{
 		}
 		return normalizedValue, nil
 	}},
+	{Scope: ScopeGlobal, Name: ValidatePasswordEnable, Value: Off, Type: TypeBool},
+	{Scope: ScopeGlobal, Name: ValidatePasswordPolicy, Value: "MEDIUM", Type: TypeEnum, PossibleValues: []string{"LOW", "MEDIUM", "STRONG"}},
+	{Scope: ScopeGlobal, Name: ValidatePasswordCheckUserName, Value: On, Type: TypeBool},
+	{Scope: ScopeGlobal, Name: ValidatePasswordLength, Value: "8", Type: TypeInt, MinValue: 0, MaxValue: math.MaxInt32,
```
[the remark this answers is not in the document]
Comment (CbcWestwolf) [reply sessionctx/variable/sysvar.go:492]: Yes, it is the same as the value in MySQL. We don't limit the length of the password.

File: sessionctx/variable/noop.go [hunk 13]
```diff
@@ -117,7 +115,6 @@ var noopSysVars = []*SysVar{
 	{Scope: ScopeNone, Name: "innodb_log_group_home_dir", Value: "./"},
 	{Scope: ScopeNone, Name: "performance_schema_events_statements_history_size", Value: "10"},
 	{Scope: ScopeGlobal, Name: GeneralLog, Value: Off, Type: TypeBool},
-	{Scope: ScopeGlobal, Name: "validate_password_dictionary_file", Value: ""},
```
Comment (xiongjiwei) [inline sessionctx/variable/noop.go:120]: this variable is still noop in tidb

File: sessionctx/variable/sysvar.go [hunk 14]
```diff
@@ -486,6 +486,86 @@ var defaultSysVars = []*SysVar{
...
+			if minLength := int32(numberCount) + specialCharCount + 2*mixedCaseCount; length < minLength {
+				PasswordValidationLength.Store(minLength)
+			}
+			return normalizedValue, nil
+		},
+		SetGlobal: func(_ context.Context, s *SessionVars, val string) error {
+			PasswordValidtaionNumberCount.Store(int32(TidbOptInt64(val, 1)))
+			return nil
+		}, GetGlobal: func(_ context.Context, s *SessionVars) (string, error) {
+			return fmt.Sprintf("%d", PasswordValidtaionNumberCount.Load()), nil
```
Comment (xiongjiwei) [inline sessionctx/variable/sysvar.go:546]: `strconv.FormatInt(PasswordValidtaionNumberCount.Load(), 10)` has better performance

Comment (CbcWestwolf) [reply sessionctx/variable/sysvar.go]: Modified :-)

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**